### PR TITLE
IsCovariant, IsContravariant.

### PR DESCRIPTION
**Pull request by TomasMikula**
See #1450.

### CODE REVIEW

File: base/shared/src/main/scala/scalaz/typeclass/IsContravariantClass.scala [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package scalaz
+package typeclass
+
+import Liskov.<~<
+
+trait IsContravariantClass[F[_]] extends IsContravariant[F] {
+  final def isCovariant: IsContravariant[F] = this
```
Comment (S11001001) [inline base/shared/src/main/scala/scalaz/typeclass/IsContravariantClass.scala]: `isContravariant`?

File: base/shared/src/main/scala/scalaz/typeclass/LiskovInstances.scala [hunk 2]
```diff
@@ -7,9 +7,6 @@ import Liskov.<~<
 
 trait LiskovInstances {
   /**We can witness equality by using it to convert between types */
-  implicit def witness[A, B](lt: A <~< B): A => B = {
-    type f[-X] = X => B
-    lt.subst[f](identity)
-  }
+  implicit def witness[A, B](lt: A <~< B): A => B = lt(_)
```
Comment (S11001001) [inline base/shared/src/main/scala/scalaz/typeclass/LiskovInstances.scala]: Hmm. If you sign this `implicit def witness[A, B](a: A)(implicit lt: A <~< B): B = lt(a)` instead, does this compile?

```scala
def qqq[A, B](implicit lt: A <~< B): A => B = a => a
```

At the very least, an `implicit` keyword is missing here.

File: base/shared/src/main/scala/scalaz/typeclass/IsContravariantClass.scala (hunk 1 above)
Comment (TomasMikula) [reply base/shared/src/main/scala/scalaz/typeclass/IsContravariantClass.scala]: Yes, thanks!

File: base/shared/src/main/scala/scalaz/typeclass/LiskovInstances.scala (hunk 2 above)
Comment (TomasMikula) [reply base/shared/src/main/scala/scalaz/typeclass/LiskovInstances.scala]: Don't know, will try.

> At the very least, an implicit keyword is missing here.

I guess @aloiscochard meant this as an implicit conversion of `<~<` to function, not to provide an implicit functions.

Comment (TomasMikula) [reply base/shared/src/main/scala/scalaz/typeclass/LiskovInstances.scala]: It compiles. That might make more sense. Note, however, that I didn't touch the signature, just the implementation.

Comment (S11001001) [reply base/shared/src/main/scala/scalaz/typeclass/LiskovInstances.scala]: To a certain degree, those are the same thing now. I'm just not sure to what degree.

Comment (TomasMikula) [reply base/shared/src/main/scala/scalaz/typeclass/LiskovInstances.scala]: @aloiscochard What do you think about changing `witness` to what @S11001001 proposed, i.e.

```scala
implicit def witness[A, B](a: A)(implicit lt: A <~< B): B = lt(a)
```

Passing `A <~< B` to a function that requires `A => B` could be achieved more simply by `Liskov` extending `Function1`.

**Pull request merged**